File: TH5_02/app/src/main/java/com/example/th5_02/MainActivity.java
package com.example.th5_02;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.telephony.PhoneStateListener;
import android.telephony.SmsManager;
import android.telephony.TelephonyCallback;
import android.telephony.TelephonyManager;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {

    private static final int REQUEST_CODE_PERMISSIONS = 100;
    private TextView callStatusTextView; // TextView để hiển thị trạng thái cuộc gọi
    private TelephonyManager telephonyManager;
    private String incomingNumber = null; // Lưu số điện thoại của cuộc gọi đến
    private boolean wasRinging = false; // Biến để theo dõi trạng thái đổ chuông
    private CallStateListener callStateListener; // Listener cho TelephonyCallback

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Ánh xạ TextView từ layout
        callStatusTextView = findViewById(R.id.call_status_textview);

        // Yêu cầu quyền runtime
        requestPermissions();
    }

    private void requestPermissions() {
        String[] permissions = {
                Manifest.permission.READ_PHONE_STATE,
                Manifest.permission.SEND_SMS
        };

        boolean needRequest = false;
        for (String permission : permissions) {
            if (ContextCompat.checkSelfPermission(this, permission)
                    != PackageManager.PERMISSION_GRANTED) {
                needRequest = true;
                break;
            }
        }

        if (needRequest) {
            ActivityCompat.requestPermissions(this, permissions, REQUEST_CODE_PERMISSIONS);
        } else {
            // Quyền đã được cấp, bắt đầu lắng nghe trạng thái cuộc gọi
            setupPhoneStateListener();
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            boolean allPermissionsGranted = true;
            for (int result : grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    allPermissionsGranted = false;
                    break;
                }
            }

            if (allPermissionsGranted) {
                Toast.makeText(this, "Tất cả quyền đã được cấp!", Toast.LENGTH_SHORT).show();
                // Quyền đã được cấp, bắt đầu lắng nghe trạng thái cuộc gọi
                setupPhoneStateListener();
            } else {
                Toast.makeText(this, "Một số quyền bị từ chối. Ứng dụng có thể không hoạt động đúng.", Toast.LENGTH_LONG).show();
            }
        }
    }

    private void setupPhoneStateListener() {
        telephonyManager = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) { // Android 12 (API 31) trở lên
            callStateListener = new CallStateListener();
            Executor executor = Executors.newSingleThreadExecutor();
            telephonyManager.registerTelephonyCallback(executor, callStateListener);
        } else {
            // Hỗ trợ cho các phiên bản Android cũ hơn (dưới API 31)
            // Sử dụng PhoneStateListener (deprecated)
            setupPhoneStateListenerLegacy();
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.S)
    private class CallStateListener extends TelephonyCallback implements TelephonyCallback.CallStateListener {
        @Override
        public void onCallStateChanged(int state) {
            switch (state) {
                case TelephonyManager.CALL_STATE_IDLE:
                    // Không có cuộc gọi
                    callStatusTextView.setText("Không có cuộc gọi");
                    // Kiểm tra nếu trước đó đang đổ chuông nhưng không trả lời (cuộc gọi nhỡ)
                    if (wasRinging && incomingNumber != null && !incomingNumber.isEmpty()) {
                        sendMissedCallMessage(incomingNumber);
                        wasRinging = false; // Đặt lại trạng thái
                        incomingNumber = null; // Xóa số điện thoại
                    } else {
                        // Nếu không có số điện thoại, thông báo lỗi
                        if (wasRinging) {
                            Toast.makeText(MainActivity.this, "Không thể gửi SMS: Không lấy được số điện thoại.", Toast.LENGTH_LONG).show();
                            wasRinging = false;
                            incomingNumber = null;
                        }
                    }
                    break;
                case TelephonyManager.CALL_STATE_RINGING:
                    // Có cuộc gọi đến
                    // Lưu ý: TelephonyCallback không cung cấp số điện thoại trực tiếp
                    // Bạn cần sử dụng cách khác để lấy số (ví dụ: BroadcastReceiver với Intent.ACTION_PHONE_STATE_CHANGED)
                    // Để đơn giản, tôi sẽ giả định số điện thoại được lấy từ cách khác
                    // Ở đây, tôi để trống incomingNumber và sẽ giải thích cách lấy số ở phần sau
                    wasRinging = true; // Đánh dấu rằng đang có cuộc gọi đến
                    callStatusTextView.setText("Có cuộc gọi đến");
                    Toast.makeText(MainActivity.this, "Có cuộc gọi đến", Toast.LENGTH_LONG).show();
                    break;
                case TelephonyManager.CALL_STATE_OFFHOOK:
                    // Đang trong cuộc gọi (gọi đi hoặc nhận cuộc gọi)
                    callStatusTextView.setText("Đang trong cuộc gọi");
                    wasRinging = false; // Cuộc gọi đã được trả lời, không phải cuộc gọi nhỡ
                    incomingNumber = null; // Xóa số điện thoại
                    break;
            }
        }
    }

    @SuppressWarnings("deprecation")
    private void setupPhoneStateListenerLegacy() {
        telephonyManager = (TelephonyManager) getSystemService(Context.TELEPHONY_SERVICE);
        PhoneStateListener phoneStateListener = new PhoneStateListener() {
            @Override
            public void onCallStateChanged(int state, String phoneNumber) {
                switch (state) {
                    case TelephonyManager.CALL_STATE_IDLE:
                        // Không có cuộc gọi
                        callStatusTextView.setText("Không có cuộc gọi");
                        // Kiểm tra nếu trước đó đang đổ chuông nhưng không trả lời (cuộc gọi nhỡ)
                        if (wasRinging && incomingNumber != null && !incomingNumber.isEmpty()) {
                            sendMissedCallMessage(incomingNumber);
                            wasRinging = false; // Đặt lại trạng thái
                            incomingNumber = null; // Xóa số điện thoại
                        } else {
                            // Nếu không có số điện thoại, thông báo lỗi
                            if (wasRinging) {
                                Toast.makeText(MainActivity.this, "Không thể gửi SMS: Không lấy được số điện thoại.", Toast.LENGTH_LONG).show();
                                wasRinging = false;
                                incomingNumber = null;
                            }
                        }
                        break;
                    case TelephonyManager.CALL_STATE_RINGING:
                        // Có cuộc gọi đến
                        incomingNumber = phoneNumber; // Lưu số điện thoại
                        wasRinging = true; // Đánh dấu rằng đang có cuộc gọi đến
                        callStatusTextView.setText("Có cuộc gọi đến từ: " + (phoneNumber != null ? phoneNumber : "Số không xác định"));
                        Toast.makeText(MainActivity.this, "Có cuộc gọi đến từ: " + (phoneNumber != null ? phoneNumber : "Số không xác định"), Toast.LENGTH_LONG).show();
                        break;
                    case TelephonyManager.CALL_STATE_OFFHOOK:
                        // Đang trong cuộc gọi (gọi đi hoặc nhận cuộc gọi)
                        callStatusTextView.setText("Đang trong cuộc gọi");
                        wasRinging = false; // Cuộc gọi đã được trả lời, không phải cuộc gọi nhỡ
                        incomingNumber = null; // Xóa số điện thoại
                        break;
                }
            }
        };

        // Bắt đầu lắng nghe trạng thái cuộc gọi
        telephonyManager.listen(phoneStateListener, PhoneStateListener.LISTEN_CALL_STATE);
    }

    private void sendMissedCallMessage(String phoneNumber) {
        try {
            SmsManager smsManager = SmsManager.getDefault();
            String message = "Tôi đang bận, hãy gọi lại sau.";
            smsManager.sendTextMessage(phoneNumber, null, message, null, null);
            Toast.makeText(this, "Đã gửi SMS đến " + phoneNumber, Toast.LENGTH_LONG).show();
        } catch (SecurityException e) {
            Toast.makeText(this, "Lỗi: Quyền gửi SMS bị từ chối.", Toast.LENGTH_LONG).show();
            e.printStackTrace();
        } catch (IllegalArgumentException e) {
            Toast.makeText(this, "Lỗi: Số điện thoại không hợp lệ.", Toast.LENGTH_LONG).show();
            e.printStackTrace();
        } catch (Exception e) {
            Toast.makeText(this, "Lỗi khi gửi SMS: " + e.getMessage(), Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        // Dừng lắng nghe trạng thái cuộc gọi khi Activity bị hủy
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && telephonyManager != null && callStateListener != null) {
            telephonyManager.unregisterTelephonyCallback(callStateListener);
        }
    }
}

File: TH05_10/app/src/main/java/com/example/th05_10/MainActivity.kt
package com.example.th05_10

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.Socket
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {

    private lateinit var messageEditText: EditText
    private lateinit var ipEditText: EditText
    private lateinit var sendButton: Button
    private lateinit var receivedMessagesTextView: TextView
    private val serverPort = 12345 // Cổng của Server
    private var socket: Socket? = null
    private var isRunning = false // Biến để kiểm soát luồng nhận tin nhắn
    private val mainHandler = Handler(Looper.getMainLooper()) // Handler để cập nhật UI
    private var writer: PrintWriter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        messageEditText = findViewById(R.id.message_edittext)
        ipEditText = findViewById(R.id.ip_edittext)
        sendButton = findViewById(R.id.send_button)
        receivedMessagesTextView = findViewById(R.id.received_messages_textview)

        sendButton.setOnClickListener {
            val message = messageEditText.text.toString().trim()
            val ipAddress = ipEditText.text.toString().trim()

            if (message.isEmpty()) {
                Toast.makeText(this, "Vui lòng nhập tin nhắn!", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (ipAddress.isEmpty()) {
                Toast.makeText(this, "Vui lòng nhập địa chỉ IP!", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }


            thread {
                connectToServer(ipAddress)
                sendMessage(message)
            }


            messageEditText.setText("")
        }
    }

    private fun connectToServer(ipAddress: String) {
        try {

            if (socket == null || socket?.isClosed == true) {
                socket = Socket(ipAddress, serverPort)
                isRunning = true
                writer = PrintWriter(socket!!.getOutputStream(), true)


                startReceivingMessages()
                mainHandler.post {
                    Toast.makeText(this, "Đã kết nối đến Server!", Toast.LENGTH_SHORT).show()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            mainHandler.post {
                Toast.makeText(this, "Lỗi khi kết nối đến Server: ${e.message}", Toast.LENGTH_LONG).show()
            }
            isRunning = false
        }
    }

    private fun sendMessage(message: String) {
        try {

            writer?.println(message)
            writer?.flush()

            mainHandler.post {
                Toast.makeText(this, "Đã gửi tin nhắn!", Toast.LENGTH_SHORT).show()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            mainHandler.post {
                Toast.makeText(this, "Lỗi khi gửi tin nhắn: ${e.message}", Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun startReceivingMessages() {
        thread {
            try {
                val reader = BufferedReader(InputStreamReader(socket?.getInputStream()))
                while (isRunning) {
                    val message = reader.readLine()
                    if (message == null) break
                    mainHandler.post {
                        val currentText = receivedMessagesTextView.text.toString()
                        receivedMessagesTextView.text = "$currentText$message\n"
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                if (isRunning) {
                    mainHandler.post {
                        Toast.makeText(this, "Lỗi khi nhận tin nhắn: ${e.message}", Toast.LENGTH_LONG).show()
                    }
                }
            } finally {
                closeConnection()
            }
        }
    }

    private fun closeConnection() {
        try {
            isRunning = false
            writer?.close()
            socket?.close()
            socket = null
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        closeConnection()
    }
}